### PR TITLE
Nette3

### PR DESCRIPTION
**Pull request by kravcik**
Todos:

- [x] `preg_match_all('/(?m)@(\S+) (\S+) (\S+(?: -> \S+)*)(.+)*$/', ($class::getReflection())->getDocComment(), $matches, PREG_SET_ORDER, 0);`
- [ ] Set up Travis for PHP7
- [x] Code review

### CODE REVIEW

**Comment by kravcik** [comment]
@uestla: Can you take a look if this is right way? :)

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
 /.idea
+/nbproject
```
Comment (uestla) [inline .gitignore]: Please remove

File: composer.json [hunk 2]
```diff
@@ -11,17 +11,16 @@
 	},
 
 	"require": {
-		"php": ">=5.6.0",
-		"nette/utils": "^2.4",
-		"nette/database": "^2.4",
-		"nette/reflection": "^2.3"
+		"php": ">=7.1.0",
+		"nette/utils": "^3",
```
Comment (uestla) [inline composer.json]: ^3.0 is better than just ^3

File: src/YetORM/Reflection/EntityType.php [hunk 3]
```diff
@@ -120,83 +117,110 @@ private function getClassTree()
...
 	{
 		if (!isset(self::$annProps[$class])) {
 			self::$annProps[$class] = [];
-			$ref = $class::getReflection();
-
-			foreach ($ref->getAnnotations() as $ann => $values) {
-				if ($ann === 'property' || $ann === 'property-read') {
-					foreach ($values as $line) {
-						$matches = NStrings::match($line, '#^[ \t]*(?P<type>\\\\?[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*(?:\\\\[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*)*(?:\|\\\\?[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*(?:\\\\[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*)*)?)[ \t]+(?P<property>[\$a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*)(?:[ \t]+->[ \t]+(?P<column>[a-zA-Z0-9_-]+))?[ \t]*(?P<description>.*)\z#');
+			preg_match_all('/(?m)@(\S+) (\S+) (\S+(?: -> \S+)*)(.+)*$/', ($class::getReflection())->getDocComment(), $matches, PREG_SET_ORDER, 0);
```
Comment (uestla) [inline src/YetORM/Reflection/EntityType.php]: Why still not use `NStrings::matchAll()` ?

File: src/YetORM/Repository.php [hunk 4]
```diff
@@ -12,12 +12,13 @@
 
 use Nette;
 use Nette\Database\IRow as NIRow;
-use Nette\Reflection\AnnotationsParser;
+use Nette\Utils\Reflection as NReflection;
```
Comment (uestla) [inline src/YetORM/Repository.php]: Please sort use statements by length.

File: tests/bootstrap.php [hunk 5]
```diff
@@ -17,6 +17,6 @@ function dd() { call_user_func_array('dump', func_get_args()); die(); }
 ]);
 
 $loader = new Nette\Loaders\RobotLoader;
-$loader->setCacheStorage(ServiceLocator::getCacheStorage());
+$loader->setTempDirectory(__DIR__ . '/temp');
```
Comment (uestla) [inline tests/bootstrap.php]: Please refactor to `ServiceLocator::getTempDir()` and use it here and in `ServiceLocator::getCacheStorage()`.

File: tests/model/ServiceLocator.php [hunk 6]
```diff
@@ -38,10 +38,25 @@ public static function getCacheStorage()
 
 
 	/** @return Nette\Database\Context */
-	public static function getDbContext()
+	public static function getDbContext(): Nette\Database\Context
 	{
 		if (self::$dbContext === NULL) {
-			$connection = new Nette\Database\Connection('mysql:host=127.0.0.1;dbname=yetorm_test', 'root', '');
+			$configFile = __DIR__ . '/../config/local.neon';
```
Comment (uestla) [inline tests/model/ServiceLocator.php]: Please remove this if/else logic since it has nothing to do with this PR.

**Review by uestla — CHANGES_REQUESTED** [review]
@kravcik Hi, please take a look at my notes & after working them in, fixup all commits into single one.

File: src/YetORM/Reflection/EntityType.php [hunk 7]
```diff
@@ -120,83 +117,110 @@ private function getClassTree()
...
+	/**
+	 * Returns an annotation value.
+	 */
+	public static function parseAnnotation(\Reflector $ref, string $name): ?string
+	{
+		if (!NReflection::areCommentsAvailable()) {
+			throw new Nette\InvalidStateException('You have to enable phpDoc comments in opcode cache.');
+		}
+		$re = '#[\s*]@' . preg_quote($name, '#') . '(?=\s|$)(?:[ \t]+([^@\s]\S*))?#';
+		if ($ref->getDocComment() && preg_match($re, trim($ref->getDocComment(), '/*'), $m)) {
```
Comment (uestla) [inline src/YetORM/Reflection/EntityType.php]: Why not use `NStrings::match()`?

**Review by uestla — CHANGES_REQUESTED** [review]

File: src/YetORM/Reflection/EntityType.php (hunk 3 above)
Comment (kravcik) [reply src/YetORM/Reflection/EntityType.php]: Yeah, good idea. We wanna fix small bug with `*/` on the end of line - now should be fixed and set to NStrings.

File: tests/model/ServiceLocator.php (hunk 6 above)
Comment (kravcik) [reply tests/model/ServiceLocator.php]: Do you think its good idea to send separete PR for this? Its hard set up automatic testing when local DB has password. Can be solved be optional config.neon.

Comment (kravcik) [reply tests/model/ServiceLocator.php]: Otherwise resolved.

**Comment by kravcik** [comment]
Should be resolved. Can you help with Travis set up for PHP7? 

There is no option for you - to squash commits when you accept PR?

**Review by uestla — CHANGES_REQUESTED** [review]
Great, thank you so much! I've just added 2 little requests, then I'll be able to merge it.

And you're right, there's the "squash and merge" button, haven't used it yet, thanks for the tip!

Hopefully I'll look at Travis.

**Comment by kravcik** [comment]
@uestla: Should prepared.

Maybe would be good idea add something like this:

https://github.com/Kdyby/FormsReplicator#versions

**Comment by uestla** [comment]
@kravcik Fantastic, thanks a lot. I'll look into it, maybe remove the rest of phpDocs & released new version.

**Comment by kravcik** [comment]
Hello, can we get at least RC tag? To start testing migrate to Nette 3?

**Comment by uestla** [comment]
To be honest there's more that hasn't been done in regards to the upgrade, I have the repo in mid-finished state and would like to finish it before merging this PR and tagging the result.

- removing useless phpDocs
- adding missing typehints
- fixing coding standards
- fixing errors found by phpstan

**Comment by kravcik** [comment]
If you need some help, let me know

File: src/YetORM/Reflection/EntityType.php [hunk 8]
```diff
@@ -76,16 +69,16 @@ private function loadEntityProperties()
...
-					&& !$method->hasAnnotation('internal')) {
+					&& !NStrings::contains($method->getDocComment(), '@internal')) {
 
 				$name = lcfirst(substr($method->getName(), 3));
-				$type = $method->getAnnotation('return');
+
+				$type = $method->getReturnType();
 
 				if ($type !== NULL && !EntityProperty::isNativeType($type)) {
 					$type = AnnotationsParser::expandClassName($type, $this);
```
Comment (pupitooo) [inline src/YetORM/Reflection/EntityType.php]: Class 'YetORM\Reflection\AnnotationsParser' not found

Comment (pupitooo) [reply src/YetORM/Reflection/EntityType.php]: I suppose that line (and condition) is not neccessary. Is it true?

**Comment by diegosardina** [comment]
Why this PR has been closed?

**Comment by kravcik** [comment]
Hey @diegosardina, this PR takes almost year and it doesnt seems it can make at least to RC. So we decided for faster variant and overhaul this repo to some beta-version by ourself. You can find it here (https://github.com/modul-is/orm) - maybe it will help you.

**Pull request closed without merging**